lib/promoted/ruby/client/request_builder.rb

```diff
@@ -125,7 +125,12 @@ def compact_one_insertion(insertion, compact_func)
 
           # Only need a copy if there are properties to compact.
           compact_insertion = insertion.dup
-          compact_insertion[:properties] = compact_func.call(insertion[:properties])
+
+          # Let the custom function work with a deep copy of the properties.
+          # There's really no way to work with a shallow copy and still be able
+          # to restore the correct insertion properties after a call to delivery.
+          new_props =  Marshal.load(Marshal.dump(insertion[:properties]))
+          compact_insertion[:properties] = compact_func.call(new_props)
```

Comment:
We should chat in person about this.  One of our Typescript client callers passes a more complex object onto properties.  I don't know if simple deep copy will work for it.

Reply:
I will enter an issue to revisit this for TS.